Removed the first recording button from SongEditor as it has no use

steven-jaro opened the pull request:
Changes:

1. Editor.h:
* Changed the name of the default value of Editor.cpp constructor from "record_step" to "recordStep".
* Added bool "recordAccompany" and set to false to reflect the changes in the parammeters on Editor.cpp.

2. Song.h:
* Removed the "record()" method as it will not be used.
* Remove parts of the code that had spaces instead of tabs size 4.

3. SongEditor.h:
* Removed the "record()" method as it will not be used.

4. Song.cpp:
* Removed the "record()" method as it will not be used.
* Remove parts of the code that had spaces instead of tabs size 4.

5. Editor.cpp:
* Added bool "recordAccompany" to the constructor.
* Renamed "stepRecord" to "recordStep" so that every parammeter name starts with "record*"
* Added a lambda functionally for conditional button creation.
* Changed the order in which the buttons are created. They were ordered in phases, now it is ordered by button.

6. PianoRoll.cpp:
* Added "(true, true, true)" to the parammeters entry for the Editor constructor to reflect which buttons need to be shown.

7. SongEditor.cpp (the key change):
* Added "(false, true, false)" to the parammeters entry for the Editor constructor to reflect which buttons need to be shown. This is for showing only the recordAccompany button in the song editor.
* Removed connection fro "m_recordAction" as it wil not be used here.
* Removed checking for "m_recordAction" as it doesn't exist anymore.
*  Removed the "record()" method as it will not be used.

How to test?:
This PR just removes the button in the Song Editor. So one only has to make sure that the correct buttons are shown on each editor and that they work properly as intended.

All of these was decided because there is no really a use for a button that records without playing the song. So the only audio recording button left is the "RecordAccompany". Also this PR solves the first issue out of 4 in #7786.

steven-jaro committed:
Removed the first record button from the Song Editor

steven-jaro committed:
Removed the first record button from the Song Editor

steven-jaro committed:
Removed commented lines in SongEditor

steven-jaro committed:
Removed commented lines in SongEditor

regulus79 commented on src/gui/editors/Editor.cpp:
```diff
@@ -89,7 +89,7 @@ void Editor::toggleMaximize()
 }
 
-Editor::Editor(bool record, bool stepRecord) :
+Editor::Editor(bool record, bool stepRecord, bool onlyRecordAccompany) :
```
Should `onlyRecordAccompany` be named something different, since currently when it's true it has both normal record and record accompany?

Or maybe it would make more sense to have bools like:
`Editor::Editor(bool record, bool recordAccompany, bool stepRecord)`

And so then if `record` is true, it adds the normal record button, if `recordAccompany` is true, adds the record accompany button, etc.

steven-jaro replied:
Oops my bad, that is actually a dumb mistake by me.

regulus79 commented on src/gui/editors/SongEditor.cpp:
```diff
@@ -1033,9 +1032,8 @@ SongEditorWindow::SongEditorWindow(Song* song) :
 	if (!Engine::audioEngine()->captureDeviceAvailable())
 	{
-		for (auto &recordAction : {m_recordAccompanyAction, m_recordAction})
+		for (auto &recordAction : {m_recordAccompanyAction})
```
I think you can remove the loop and just directly access `m_recordAccompanyAction` since it's the only one.

steven-jaro replied:
Understood.

regulus79 commented on src/gui/editors/SongEditor.cpp:
```diff
@@ -1099,12 +1097,6 @@ void SongEditorWindow::play()
 }
 
 
-void SongEditorWindow::record()
-{
-	m_editor->m_song->record();
-}
-
-
```
Now I'm wondering... maybe this isn't relevant to this PR, but now that the song editor only has one record button, would it make sense to just call it the record button, instead of record accompany? You don't have to do that in this PR I'm just thinking out loud

steven-jaro replied:
I don't blame you, I also thought about that. I actually had that idea but it would actually be overcomplicated for the following reasons:
1. We are conditioning the normal record button to appear, not the step record, so we would have to make the step record as conditional.
2. We would have to pass the functionality from the step record to the normal record button.
3. I think the name of the function recordAccompany is very good as it nicely describes what it does.

If more people want this, I can make it, I don't have any problem. I am not so into the idea but in the long road I am indifferent to that change.

steven-jaro committed:
Fixed regulus request about inverted bool name and removed an unnecessary loop

steven-jaro committed:
Fixed regulus request about inverted bool name and removed an unnecessary loop

steven-jaro commented:
I added the changes you asked me.

regulus79 commented on src/gui/editors/Editor.cpp:
```diff
@@ -126,7 +132,10 @@ Editor::Editor(bool record, bool stepRecord) :
 	addButton(m_playAction, "playButton");
 	if (record)
 	{
-		addButton(m_recordAction, "recordButton");
+		if (recordAccompany)
+		{
+			addButton(m_recordAction, "recordButton");
+		}	
 		addButton(m_recordAccompanyAction, "recordAccompanyButton");
 	}
```
Wait so.... if `recordAccompany` is true, it adds the `record` button.

But either way it will still add the `recordAccompany` button?

I'm wondering if it would make more sense if it were like:

```cpp
if (record)
{
	addButton(m_recordAction, "recordButton");
}
if (recordAccompany)
{
	addButton(m_recordAccompanyAction, "recordAccompanyButton");
}
```

steven-jaro committed:
Modified editor buttons construction logic

steven-jaro committed:
Modified editor buttons construction logic

steven-jaro commented:
I modified the logic of the set up for all the buttons so when adding the conditions, the code looks as compact and simplified as possible. The code order was:
[Action set up for each button (play, record, recordAccompany, stepRecord, Stop)] ->
[Connection set up for each button (play, record, recordAccompany, stepRecord, Stop)] ->
[Toolbar set up for each button (play, record, recordAccompany, stepRecord, Stop)]

And now I changed it to:
[Complete setup for play (action, connection, toolbar)] ->
[Complete setup for record (action, connection, toolbar)] ->
[Complete setup for recordAccompany (action, connection, toolbar)] ->
[Complete setup for stepRecord (action, connection, toolbar)] ->
[Complete setup for stop (action, connection, toolbar)] ->

I made this change because otherwise I would have to add 9 "ifs" if I wanted to keep the previous order. So for adding just 3 "ifs" I reordered the logic of the code grouping by button instead of by set up stage. I also changed the order of the bools in the constructor and updated the change for the constructor of the PianoRoll and SongEditor

regulus79 commented:
Nice! That does look much cleaner

regulus79 commented on include/Editor.h:
```diff
@@ -80,7 +80,7 @@ private slots:
 	/// \param	record	If set true, the editor's toolbar will contain record
 	///					buttons in addition to the play and stop buttons.
-	Editor(bool record = false, bool record_step = false);
+	Editor(bool record = false, bool recordAccompany = true, bool record_step = false);
```
I believe this should be false by default, since currently I think it's causing the Pattern Editor and Automation Editor to have the record accomany button, and I don't think they had them before.

![image](https://github.com/user-attachments/assets/0e3bedf8-d4d3-4c0b-84ac-0ac0dbea2835)

steven-jaro replied:
You are right. Let me solve that quickly.

steven-jaro committed:
Fixed recordAccompany as false by default

steven-jaro committed:
Fixed recordAccompany as false by default

steven-jaro commented:
The error is now fixed.

szeli1 commented on src/gui/editors/Editor.cpp:
```diff
@@ -104,36 +104,44 @@ Editor::Editor(bool record, bool stepRecord) :
-	m_playAction = new QAction(embed::getIconPixmap("play"), tr("Play (Space)"), this);
-	m_stopAction = new QAction(embed::getIconPixmap("stop"), tr("Stop (Space)"), this);
+	// Play action setup
+	m_playAction = new QAction(embed::getIconPixmap("play"), tr("Play (Space)"), this); //setup play action
```
```suggestion
	m_playAction = new QAction(embed::getIconPixmap("play"), tr("Play (Space)"), this); // setup play action
```

szeli1 commented on src/gui/editors/Editor.cpp:
```diff
@@ -104,36 +104,44 @@ Editor::Editor(bool record, bool stepRecord) :
+	// RecordAccompany action setup
+	if (recordAccompany)
+	{
+		m_recordAccompanyAction = new QAction(embed::getIconPixmap("record_accompany"), tr("Record while playing"), this); //setup recordAccompany action
```
```suggestion
		m_recordAccompanyAction = new QAction(embed::getIconPixmap("record_accompany"), tr("Record while playing"), this); // setup recordAccompany action
```

szeli1 commented on src/gui/editors/Editor.cpp:
```diff
@@ -104,36 +104,44 @@ Editor::Editor(bool record, bool stepRecord) :
+	// RecordAccompany action setup
+	if (recordAccompany)
+	{
+		m_recordAccompanyAction = new QAction(embed::getIconPixmap("record_accompany"), tr("Record while playing"), this); //setup recordAccompany action
```
Also better naming, please could you replace every `recordAccompany` with a better name?

steven-jaro replied:
> Also better naming, please could you replace every `recordAccompany` with a better name?

I concluded that this should be adress in another PR as this is focussed on removing the recording button from the Song Editor.

szeli1 commented on include/Editor.h:
```diff
@@ -80,7 +80,7 @@ private slots:
 	/// \param	record	If set true, the editor's toolbar will contain record
 	///					buttons in addition to the play and stop buttons.
-	Editor(bool record = false, bool record_step = false);
+	Editor(bool record = false, bool recordAccompany = false, bool record_step = false);
```
```suggestion
	Editor(bool record = false, bool recordAccompany = false, bool record_step = false);
```
Better naming? Maybe `allowRecording`? or `isRecordingEnabled`?

steven-jaro replied:
@regulus79 and I discussed this previously. These variables are from the constructor of the editor, each name is assigned to one of the recording buttons. Do you think it is really neccesary to change the names of the variables to

Editor(bool allowRecording = false, bool allowRecordingAccompany = false, bool allowStepRecording = false);

or

Editor(bool isRecordingEnabled = false, bool isRecordingAccompanyEnabled = false, bool isStepRecordingEnabled = false);

?
Personally, I think it is unnecessary.

regulus79 replied:
It's only true in the piano roll and song editor; the automation editor and pattern editor default to false because they don't support the same kind of recording.

szeli1 commented on src/gui/editors/Editor.cpp:
```diff
@@ -104,36 +104,44 @@ Editor::Editor(bool record, bool stepRecord) :
 	{
-		addButton(m_toggleStepRecordingAction, "stepRecordButton");
+		m_toggleStepRecordingAction = new QAction(embed::getIconPixmap("record_step_off"), tr("Toggle Step Recording"), this); //setup stepRecord action
+		connect(m_toggleStepRecordingAction, SIGNAL(triggered()), this, SLOT(toggleStepRecording())); // Set up connection for stepRecord action
```
Also no need to comment every line.

szeli1 replied:
> I kept the amount of documentation as I think it is useful

This kind of gui code is everywhere, so I don't really see why it is needed to document every line using the same words like "connect" in a line that starts with `connect(`, or action in a line that has `QAction` already in it.

> they were already there

I didn't really see any deleted comments apart from 2. I like the comments before the if statements, but I think for every line its too much.

szeli1 reviewed (commented):
Please could you address my suggestions? Also I did not test this PR.

steven-jaro committed:
Fixed szeli1 request for consisten documentation

steven-jaro committed:
Fixed szeli1 request for consisten documentation

steven-jaro commented:
@szeli1 I made the changes you requested but I kept the amount of documentation as I think it is useful.

szeli1 reviewed (approved):
I will approve this, but I did not test. I still don't like how many comments there are.

szeli1 commented:
> I will adress your suggestion for refactoring some bad variable names (in another PR).

Don't bother with it, thanks for working on this issue.

steven-jaro commented:
> I will approve this, but I did not test. I still don't like how many comments there are.

I would like to know the opinion from the other devs. If another devs says it is too much commenting, then I will remove them.

steven-jaro committed:
Fixed too much comments and implementation of a lambda to simplify the implementation

steven-jaro committed:
Fixed too much comments and implementation of a lambda to simplify the implementation

szeli1 reviewed (approved):
Making a lambda was a good Idea, I couldn't find any issues, so I will approve this

JohannesLorenz commented on src/gui/editors/Editor.cpp:
```diff
@@ -104,36 +104,40 @@ Editor::Editor(bool record, bool stepRecord) :
+		const QString& objectName
+	) {
+		if (condition) {
+			targetAction = new QAction(embed::getIconPixmap(iconName), text, this);
```
Would this be a good opportunity to change the `QAction` pointers to `unique_ptr`?

JohannesLorenz commented on include/Editor.h:
```diff
@@ -80,7 +80,7 @@ private slots:
 	/// \param	record	If set true, the editor's toolbar will contain record
 	///					buttons in addition to the play and stop buttons.
-	Editor(bool record = false, bool record_step = false);
+	Editor(bool record = false, bool recordAccompany = false, bool record_step = false);
```
If you think it makes sense, change `record_step` to `recordStep` and update the doxygen comment above to reflect all 3 parameters.

JohannesLorenz reviewed (approved):
I only have 3 style suggestions left, but overall this PR is good and can go into testing phase.

steven-jaro committed:
Fixed use of new function for all buttons and step_record to recordStep

steven-jaro committed:
Fixed use of new function for all buttons and step_record to recordStep

JohannesLorenz commented:
You changed all TABs to spaces, can you please fix? Also, what about the unique_ptr question?

JohannesLorenz commented on src/gui/editors/Editor.cpp:
```diff
@@ -89,51 +89,51 @@ void Editor::toggleMaximize()
+        m_toolBar->widgetForAction(action)->setObjectName(objectName);
+    };
+
+    auto createButton = [this, &addButton](
```
`addButton` is now only ever used inside `createButton`, so instead of declaring it and passing to `createButton`, I would just copy the 2 code lines into `createButton`?

steven-jaro committed:
Fixed 4 tab size instead of 4 spaces

steven-jaro committed:
Fixed 4 tab size instead of 4 spaces

JohannesLorenz commented on include/Song.h:
```diff
@@ -238,7 +238,7 @@ class LMMS_EXPORT Song : public TrackContainer
 		m_loopRenderRemaining = m_loopRenderCount;
 	}
-    
+		
```
No whitespace needed in the whole line (`git diff` marks all trailing whitespace red, which is why we usually keep changes clean of them).

JohannesLorenz reviewed (approved):
3 Approvals, ready for testing phase.

Feel free to fix the whitespace comment.

steven-jaro committed:
Update Song.h

steven-jaro committed:
Update Song.h

steven-jaro committed:
Update Song.h

steven-jaro committed:
Update Song.h

bratpeki commented on src/core/Song.cpp:
```diff
@@ -731,7 +723,7 @@ void Song::startExport()
 	{
 		m_exportSongEnd = TimePos(m_length, 0);
-        
+		
```
Please drop these extra tabs, they're hurting me at an emotional level! :rofl:

steven-jaro replied:
Okay XD

steven-jaro committed:
Update Song.cpp

steven-jaro committed:
Update Song.cpp

bratpeki commented:
I trust @regulus79, @szeli1 and @JohannesLorenz that the PR is well made, and the button is not there, so I approve it. Still, please change the following before merging.

bratpeki commented on src/core/Song.cpp:
```diff
@@ -456,7 +456,7 @@ bool Song::isExportDone() const
 {
 	TimePos pos = getPlayPos();
-    
+	
```
.

bratpeki commented on src/core/Song.cpp:
```diff
@@ -740,7 +731,7 @@ void Song::startExport()
 		if (!m_exportLoop) 
 			m_exportSongEnd += TimePos(1,0);
-        
+		
```
.

bratpeki reviewed (commented):
Remove unneeded whitespace.

steven-jaro committed:
Merge branch 'feature/recording-stage-one' into feature/recording-stage-one

steven-jaro committed:
fixed spaces into tabs

steven-jaro committed:
Fixed conflicts with new changes

steven-jaro committed:
Merge branch 'feature/recording-stage-one' of https://github.com/steven-jaro/lmms into feature/recording-stage-one

steven-jaro committed:
Showed record accompany into patterneditor and removed whitespaces

JohannesLorenz commented:
> unless anyone objects

Let's wait until at least the spacing is fixed ;)

Rossmaxx commented:
Sure @JohannesLorenz

bratpeki commented on src/gui/editors/PatternEditor.cpp:
```diff
@@ -223,7 +223,7 @@ void PatternEditor::cloneClip()
 
 PatternEditorWindow::PatternEditorWindow(PatternStore* ps) :
-	Editor(false),
+	Editor(false, true, false),
```
BB Editor doesn't actually record anything, right?

I know it's out of the scope of this PR, I'm asking purely informatively.

bratpeki replied:
[Segfault backtrace, log.txt](https://github.com/user-attachments/files/25135619/log.txt)

bratpeki committed:
Wording change

bratpeki committed:
Made it so you can't record from Pattern Editor

regulus79 reviewed (approved):
I believe I approved this PR a while ago, but it seems github removed my approval when I was requested to review again. I don't think any major changes were made, so I am guessing the PR should still be fine.
